appinventor/appengine/src/com/google/appinventor/client/explorer/youngandroid/ProjectList.java

```diff
@@ -319,139 +198,182 @@ public void refreshTable(boolean needToSort, boolean isInTrash) {
    */
   public int getSelectedProjectsCount() {
-    return selectedProjects.size();
+    if (folder != null) {
+      return folder.getSelectedProjects().size() + folder.getSelectedFolders().size();
+    }
+    else return 0;
   }
 
+
   public int getMyProjectsCount() {
     int count = 0;
-    for (Project project : projects) {
+    if (folder == null) {
+      return 0;
+    }
+    for (Project project : folder.getVisibleProjects()) {
       if (!project.isInTrash()) {
         ++ count;
       };
     }
     return count;
   }
 
-  /**
-   * Returns if the specified project is in Trash or in MyProjects
-   */
-  public int getProjectCurrentView(Project project) {
-    if (project.isInTrash()) {
-      return Ode.TRASHCAN;
+  public void setIsTrash(boolean isTrash) {
+    this.isTrash = isTrash;
+    if (isTrash) {
+      folder = Ode.getInstance().getFolderManager().getTrashFolder();
     } else {
-      return Ode.PROJECTS;
+      folder = Ode.getInstance().getFolderManager().getGlobalFolder();
+    }
+    if (folder != null) {
+      refresh();
     }
   }
 
-  /**
-   * Returns true if all projects under the current view have been selected, and false if not.
-   */
-  public boolean isAllProjectsSelected() {
-    if (Ode.getInstance().getCurrentView() == Ode.PROJECTS
-          && getSelectedProjectsCount() == getMyProjectsCount()) {
-      return true;
-    }
-    if (Ode.getInstance().getCurrentView() == Ode.TRASHCAN
-          && getSelectedProjectsCount() == projects.size() - getMyProjectsCount()) {
-      return true;
-    }
-    return false;
+  // FolderManagerEventListener implementation
+  @Override
+  public void onFolderAdded(ProjectFolder folder) {
+    refresh();
   }
-  /**
-   * Returns the list of selected projects.
-   *
-   * @return the selected projects
-   */
-  public List<Project> getSelectedProjects() {
-    return selectedProjects;
+
+  @Override
+  public void onFolderRemoved(ProjectFolder folder) {
+    refresh();
+  }
+
+  @Override
+  public void onFolderRenamed(ProjectFolder folder) {
+    refresh();
+  }
+
+  @Override
+  public void onFoldersChanged() {
+    refresh();
   }
 
-  //  ProjectManagerEventListener implementation
+  @Override
+  public void onFoldersLoaded() {
+    setIsTrash(isTrash);
+  }
 
   @Override
   public void onProjectAdded(Project project) {
-    projects.add(project);
-    projectWidgets.put(project, new ProjectWidgets(project));
-    if (!projectListLoading) {
-      refreshTable(true);
+    if (folder == null) {
+    }
+    if (projectsLoaded) {
+      folder.addProject(project);
+      Ode.getInstance().getFolderManager().saveAllFolders();
+      refresh();
     }
   }
 
   @Override
   public void onTrashProjectRestored(Project project) {
-    selectedProjects.remove(project);
-    refreshTable(false);
-    Ode.getInstance().getProjectToolbar().updateButtons();
+    Ode.getInstance().getFolderManager().getGlobalFolder().addProject(project);
```

Comment:
Hmm. If I move a put a project from a folder into the trash, restoring it puts it back into `*global*` rather than the original folder it was in. I'm not sure if that is intentional. If so, we may want to document that somewhere.

Reply:
> Hmm. If I move a put a project from a folder into the trash, restoring it puts it back into `*global*` rather than the original folder it was in. I'm not sure if that is intentional. If so, we may want to document that somewhere.

That is intentional. 

In Mac Finder, a trashed file isn't automatically restored to its original folder. if a project is selected individually and then trashed, it's placed in the trash can without its folder. There is a "Put Back" menu option to put it back where it came from, but I only recently found that option myself. The expected behavior is to just drag it where you want it to go. If a folder is trashed, then the folder structure and the files contained in it are placed in the Trash can, but then the entire unit has to be restored or deleted/emptied -- individual files can't be restored. This is a little less confusing because it's mostly done with drag-drop.

Moreover, if I kept track of the folder a project was in when it was trashed, I'd need to be able to handle subsequent changes to the folder structure while a project was in Trash before being restored.

It seemed like any choice was going to be confusing to someone, and I thought the nuance might cause more trouble than convenience. I went for handling all projects the same and stripping them out of folders when they were trashed. If someone trashes hundred of projects at the same time in a complex folder structure and then decides they want to restore the whole thing, they'll probably be disappointed, but this seemed less likely to surprise users in the usual case.

Reply:
Sounds reasonable.